---
title: "CD34 Transcript Isoform Expression"
author: "Jenny Smith"
date: "11/2/2021"
output: html_document
---

#Set-Up

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, 
                      fig.align='center', fig.height = 10,
                      fig.width = 10)
knitr::opts_knit$set(root.dir = file.path(PROJHOME,'/2018.02.12_AML_Specific_Transcripts/'))
options(stringsAsFactors=FALSE,  java.parameters = "-Xmx4g",
        bitmapType = 'cairo',device='x11' )
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message=FALSE}
library(tibble)
library(tidyr)
library(purrr)
library(dplyr)
library(magrittr)
library(stringr)

library(ggplot2)
library(gridExtra)

# library(ggpubr)
# library(ggridges)
# library(reshape2)
# library(patchwork)

library(DeGSEA) #need to install
library(REDCapR)

getwd()
```


# Raw Counts

```{r}
genome <- "GRCh38"
```

## GRCh38 


```{r}
current_files <- dir(file.path(PROJHOME, "0000.00.03_ExpressionMatrices/Kallisto_GRCh38_Gencode_v29/"))
# current_files
    

if(genome=="GRCh38"){
    grch38_cts_file <- grep("_RBD_.+scaledTPM_counts.RDS", current_files, value=TRUE)
    cts_grch38 <- readRDS(file.path(PROJHOME, "0000.00.03_ExpressionMatrices/Kallisto_GRCh38_Gencode_v29/",grch38_cts_file))
    cts_grch38_ids <- cts_grch38[,grep("gene_id|gene_name", colnames(cts_grch38))]
    
    cts_grch38 <- as.data.frame(cts_grch38)
    rownames(cts_grch38) <-  cts_grch38_ids$gene_name
    cts_grch38 <- cts_grch38[,-grep("gene_id|gene_name", colnames(cts_grch38))]
    
    # head(cts_grch38[,1:5])
    dim(cts_grch38) #58263  3021 
    
    ### TPM
    grch38_TPM_file <- grep("_RBD_.+Abundance_TPM", current_files, value=TRUE)
    TPM_grch38 <- readRDS(file.path(PROJHOME, "0000.00.03_ExpressionMatrices/Kallisto_GRCh38_Gencode_v29/",grch38_TPM_file))
    TPM_grch38_ids <- TPM_grch38[,grep("gene_id|gene_name", colnames(TPM_grch38))]
    
    TPM_grch38 <- as.data.frame(TPM_grch38)
    rownames(TPM_grch38) <-  TPM_grch38_ids$gene_id
    TPM_grch38 <- TPM_grch38[,-grep("gene_id|gene_name", colnames(TPM_grch38))]
    
    
    # head(TPM_grch38)
    dim(TPM_grch38) #58263  3021
}
```


```{r}
transcript.TPM.full <- readRDS(file.path(PROJHOME,"0000.00.03_ExpressionMatrices/Kallisto_GRCh38_Gencode_v29/transcript_level/TARGET_AML_RBD_Dx_Rlps_NBM_MPN_Kallisto_Quant_TranscriptLevel_Abundance_TPM.RDS"))
colnames(transcript.TPM.full)[grep("PATGIG|PATISD", colnames(transcript.TPM.full))] <- gsub("_replicate", "", grep("PATGIG|PATISD", colnames(transcript.TPM.full), value=T))



dim(transcript.TPM.full)
head(transcript.TPM.full[,1:5])
```

# Gene Annotations

```{r}
transcriptID.map <- read.delim(file.path(PROJHOME,"0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_TranscriptLevel_IDmap_1.18.20.txt"))

dim(transcriptID.map)
```

# ClinData


```{r message=FALSE}
#https://cran.r-project.org/web/packages/REDCapR/vignettes/workflow-read.html 
project <- "CD34_Transcript_Level_Expression"

if(project==""){
  stop("Must include Projecy name!")
}else{
  message(paste0("Project is: ",project))
  current_cde_database <- paste("TARGET_AML_CDEs_For_Project",project, ".RDS", sep="_")
  
  if(file.exists(current_cde_database)){
    print("Reading CDEs from Rdata object.")
    merged <- readRDS(current_cde_database)
      
  }else{
    print("Downloading CDEs from RedCap API.")
    path_credential <- file.path(HOME,".redcap")
    project_id <- 1295
  
    credential  <- REDCapR::retrieve_credential_local(
      path_credential = path_credential,
      project_id = project_id)
    
    #takes about 30 sec to download. 
    merged <- redcap_read(redcap_uri = credential$redcap_uri, 
                          token = credential$token, 
                          raw_or_label_headers = 'label')
    if(merged$success){
      merged <- data.frame(merged$data, check.names=TRUE) #remove the white spaces  (will this even work??)
      saveRDS(merged, current_cde_database)
    }
    
    #Create a simple log file from the day the project starts
    cat(c(paste("Date:", Sys.Date()),
          paste("cts:", basename(get(ls(pattern = "_cts_file")))),
          paste("tpm:", basename(get(ls(pattern = "_TPM_file")))), 
          paste("CDE:", current_cde_database)),
          sep = "\n", 
          file = paste(project, Sys.Date(), ".log", sep="_"))
    
  }
  
  #keep a list of the ineligable patiens to remove if necessary
  inelig <- merged %>% 
    filter(Eligibility.Comments == "remove") %>% 
    pull(USI)
  
  #Filter those with USIs and are eligible for the study
  merged <- merged %>% 
    filter(Eligibility.Comments != "remove")
  
  
  dim(merged) 
  head(merged)

}
```


```{r}
sample_info <- read.csv(file.path(TARGET, "SequencingDataMatrix/TARGET_AML_Ribodepleted_Manifest_08.12.21.csv")) 

dim(sample_info)
```

```{r}
CD34_classes <- openxlsx::read.xlsx("CD34_categorical_list.xlsx")

head(CD34_classes)
```

# Define Samples 

```{r}
samples <- sample_info %>% 
  filter(grepl("diagnostic|NBM", Time_point),
         grepl("^AML|NBM", Group)) %>% 
  filter(!grepl("replicate", Sample)) %>% 
  left_join(., CD34_classes, by="USI") %>% 
  mutate_at(vars(CD34.category), ~case_when(
    is.na(.) & Group=="NBM" ~ Group, 
    TRUE ~ .  )) %>% 
  left_join(., select(merged,USI, NPM.mutation., 
                      FLT3.ITD.positive.), by="USI") %>% 
  filter(!is.na(CD34.category)) 



# table(samples$AML_Subtype)
table(samples$Group)
table(samples$CD34.category)
dim(samples) #1182   20
```



# Transcript Level Expression 

```{r}
CD34_tx_ref <- transcriptID.map %>% 
  filter(gene_name=="CD34")


CD34_tx_ref
```

```{r}
CD34.Txs <- transcript.TPM.full[CD34_tx$transcript_id, colnames(transcript.TPM.full) %in% samples$Sample] %>% 
  as.data.frame() %>% 
  rownames_to_column("transcript_id") %>% 
  gather(Sample, TPM, -transcript_id) %>% 


  group_by(Sample) %>%
  mutate(Rank=rank(TPM, ties.method = "first")) %>% 
  mutate(DominantSpecies=case_when(
    sum(TPM) < 1.0 ~ "NotExpressed",
    Rank==max(Rank) ~ transcript_id[Rank==max(Rank)],
    Rank!=max(Rank) ~ as.character("NotDominant"))) %>% 
  ungroup() %>% 
  
  left_join(.,select(CD34_tx_ref,transcript_id,gene_name,gene_id),
            by=c("transcript_id")) %>% 
  left_join(., select(samples,Sample,AML_Subtype, CD34.category, NPM.mutation.), by="Sample") %>% 
  mutate_at(vars(NPM.mutation.), ~case_when(
    .=="Yes" ~ "NPM1+",
    .=="No" ~ "NPM1-", 
    TRUE ~ AML_Subtype))

head(CD34.Txs)
# dim(CD34.Txs)
table(CD34.Txs$NPM.mutation.)
```


```{r}
CD34.Txs.annotated <- CD34.Txs %>% 
  pivot_wider(id_cols = c(Sample,DominantSpecies, gene_name, gene_id),
              names_from=transcript_id,
              values_from=TPM) %>% 
  group_by(Sample) %>%
  mutate_at(vars(matches("^ENST")), ~replace(.,is.na(.),0)) %>%
  mutate_at(vars(matches("^ENST")), sum) %>%
  ungroup() %>% 


  #remove duplicated rows
  filter(grepl("ENST|NotExpressed", DominantSpecies)) %>%
  group_by(Sample) %>%
  mutate_at(vars(DominantSpecies), ~case_when(
    n() == 2 ~ .[.!="NotDominant"],
    TRUE ~ .)) %>%
  ungroup() %>% 
  filter(!duplicated(Sample)) %>% 

  left_join(., select(samples,Sample,USI,Protocol,AML_Subtype,
                      Primary.Fusion,Group,Tissue, CD34.category),
            by="Sample") %>%
  left_join(.,select(CD34_tx_ref,transcript_id,
                     Biotype_Dominant_Transcript=gene_type,
                     Protein_id_Dominant_Transcript=protein_id),
            by=c("DominantSpecies"="transcript_id")) %>%
  rename_at(vars(matches("^ENST")),~paste0(., "_TPM")) %>% 
  arrange(AML_Subtype,CD34.category, DominantSpecies) %>% 
  select(Sample,
         USI:CD34.category,
         gene_name,
         gene_id,
         DominantSpecies,
         Protein_id_Dominant_Transcript,
         matches("Biotype"),
         matches("^ENST"),
         everything())


CD34.Txs.annotated
# dim(CD34.Txs.annotated) 
```

```{r}
# table(CD34.Txs.annotated$CD34.category, useNA = "always")
```

```{r}
# write.csv(CD34.Txs.annotated,"TARGET_AML_CD34_Expressors_by_FlowCytometry_GRCh38_transcript_expression.csv", row.names = FALSE)
```

# Expression Plots

```{r}
# head(CD34.Txs)
CD34_tx_ref
```

	
collapsing to unique 'x' values
collapsing to unique 'x' values
collapsing to unique 'x' values
collapsing to unique 'x' values

```{r}
library(ggpubr)
```

```{r}
?get_palette
```


```{r fig.width=12, fig.height=8, warning=FALSE}
AML_allcomers <- ggplot(CD34.Txs,aes(x=CD34.category,y=TPM, fill=CD34.category)) +
  # geom_boxplot()
  geom_point(aes(color=CD34.category), size=0.5, position = position_jitterdodge()) +
  geom_violin(scale = "width", draw_quantiles = 0.5, alpha=0.3) +
  facet_wrap(~transcript_id, scale="free_x") +
  theme_classic() +
  scale_fill_manual(values=get_palette("jco", 3)) +
  scale_color_manual(values=get_palette("jco", 3))
  

```


```{r}
# ggsave(filename="TARGET_AML_CD34_transcript_expression_violin_plot.pdf", plot=AML_allcomers, device = "pdf", height = 8, width = 12)
```

```{r fig.width=12, fig.height=8, warning=FALSE}
NPM1 <- ggplot(filter(CD34.Txs,NPM.mutation.=="NPM1+"),
       aes(x=CD34.category,y=TPM, fill=CD34.category)) +
  # geom_boxplot()
  geom_point(aes(color=CD34.category), size=0.5, position = position_jitterdodge()) +
  geom_violin(scale = "width", draw_quantiles = 0.5, alpha=0.3) +
  facet_wrap(~transcript_id+NPM.mutation., scale="free_x") +
  theme_classic() +
  scale_fill_manual(values=get_palette("jco", 3)) +
  scale_color_manual(values=get_palette("jco", 3))

NPM1

ggsave(file="TARGET_AML_CD34_transcript_expression_NPM1_violin_plot.pdf", plot=NPM1, device = "pdf", height = 8, width = 12)
```

```{r fig.width=12, fig.height=8, warning=FALSE}
KMT2A <- ggplot(filter(CD34.Txs,AML_Subtype=="KMT2A"),
       aes(x=CD34.category,y=TPM, fill=CD34.category)) +
  # geom_boxplot()
  geom_point(aes(color=CD34.category), size=0.5, position = position_jitterdodge()) +
  geom_violin(scale = "width", draw_quantiles = 0.5, alpha=0.3) +
  facet_wrap(~transcript_id+AML_Subtype, scale="free_x") +
  theme_classic() +
  scale_fill_manual(values=get_palette("jco", 3)) +
  scale_color_manual(values=get_palette("jco", 3))

KMT2A
# ggsave(filename = "TARGET_AML_CD34_transcript_expression_KMT2A_violin_plot.pdf", plot=KMT2A,device = "pdf", height = 8, width = 12)
```


# Session Information

```{r}
sessionInfo()
```



